modules/xml/src/smithy4s/xml/internals/XmlSchemaVisitor.scala

```diff
@@ -0,0 +1,138 @@
+            nodes.zipWithIndex
+              .traverse { case (elem, index) =>
+                memberReader.read(
+                  XmlCursor.SingleNode(history.appendIndex(index), elem)
+                )
+              }
+              .map(list => tag.fromIterator(list.iterator))
+          case XmlCursor.SingleNode(history, elem) =>
+            memberReader
+              .read(XmlCursor.SingleNode(history.appendIndex(0), elem))
+              .map(value => tag.fromIterator(Iterator.single(value)))
+          case XmlCursor.NoNode(_) => Right(tag.empty)
+          case other =>
+            Left(XmlReadError(other.history, s"Expected one or multiple nodes"))
+        }
+      }
+    }
+  }
+
+  def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): XmlDecoder[Map[K, V]] = ???
+
+  def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): XmlDecoder[E] = ???
+
+  def struct[S](
+      shapeId: ShapeId,
+      hints: Hints,
+      fields: Vector[SchemaField[S, _]],
+      make: IndexedSeq[Any] => S
+  ): XmlDecoder[S] = {
+    def fieldReader[A](field: SchemaField[S, A]): XmlDecoder[A] = {
+      val isAttribute = field.instance.hints.has(XmlAttribute)
+      val xmlName =
+        field.instance.hints.get(XmlName).map(_.value).getOrElse(field.label)
+      field
+        .foldK(new Field.FolderK[Schema, S, XmlDecoder] {
+          def onRequired[AA](
+              label: String,
+              instance: Schema[AA],
+              get: S => AA
+          ): XmlDecoder[AA] = {
+            if (isAttribute) compile(instance).attribute(xmlName)
+            else compile(instance).down(xmlName)
+          }
+          def onOptional[AA](
+              label: String,
+              instance: Schema[AA],
+              get: S => Option[AA]
+          ): XmlDecoder[Option[AA]] = {
+            if (isAttribute) compile(instance).optional.attribute(xmlName)
+            else compile(instance).optional.down(xmlName)
+          }
+        })
+    }
+    val readers = fields.map(fieldReader(_))
+    new XmlDecoder[S] {
+      def read(cursor: XmlCursor): Either[XmlReadError, S] =
+        readers.traverse(_.read(cursor)).map(make)
+    }
+  }
+
+  def union[U](
+      shapeId: ShapeId,
+      hints: Hints,
+      alternatives: Vector[SchemaAlt[U, _]],
+      dispatch: Alt.Dispatcher[Schema, U]
+  ): XmlDecoder[U] = ???
+
+  def biject[A, B](
+      schema: Schema[A],
+      bijection: Bijection[A, B]
+  ): XmlDecoder[B] =
+    schema.compile(this).map(bijection.to)
+
+  def refine[A, B](
+      schema: Schema[A],
+      refinement: Refinement[A, B]
+  ): XmlDecoder[B] =
+    schema.compile(this).emap(refinement.asFunction)
+
+  def lazily[A](suspend: Lazy[Schema[A]]): XmlDecoder[A] = new XmlDecoder[A] {
+    lazy val underlying: XmlDecoder[A] = suspend.map(compile(_)).value
```

Comment:
perhaps we can avoid the double lazy (Eval has one built in) and move the `.value` to inside `read?

Reply:
Actually I had tried to remove the double lazy in the json decoder, and it was actually slightly more performant to keep way it was. I was quite surprised by it. I may be wrong, but I think it's because it's may actually hitting to lazy vals on the hotpath.

Reply:
interesting